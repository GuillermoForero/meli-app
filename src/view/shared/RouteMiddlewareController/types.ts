export interface RouteMiddlewareControllerProps {
  middlewares: React.ReactElement<MiddlewareProps>[];
  children: React.ReactElement;
}

export interface MiddlewareProps {
  next?: () => void;
}
