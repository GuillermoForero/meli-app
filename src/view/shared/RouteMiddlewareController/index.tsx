import React, { useState, useCallback } from 'react';

import { RouteMiddlewareControllerProps } from 'view/shared/RouteMiddlewareController/types';
//The implementation of a kind of middleware and middleware controller is in general
//to execute actions before the component or the view is rendered and thus when it is rendered it has what it needs loaded
const RouteMiddlewareController: React.FC<RouteMiddlewareControllerProps> = ({
  children,
  middlewares,
}: RouteMiddlewareControllerProps) => {
  const middlewaresLength = middlewares.length;

  if (!middlewaresLength) {
    throw new Error('The middlewares array prop should not be empty');
  }
  const [currentMiddleware, setCurrentMiddleware] = useState(1);

  const next = useCallback(() => {
    setCurrentMiddleware(m => m + 1);
  }, []);

  const componentToRender =
    currentMiddleware <= middlewaresLength
      ? middlewares[currentMiddleware - 1]
      : children;

  return (
    <>
      {React.cloneElement(componentToRender, {
        next: currentMiddleware <= middlewaresLength ? next : undefined,
      })}
    </>
  );
};

export default RouteMiddlewareController;
