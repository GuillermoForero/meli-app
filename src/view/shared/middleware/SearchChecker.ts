import { REQUEST_STATUS } from 'interfaces/state';
import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { ROUTES_NAVIGATION } from 'router/routes';
import { itemsActions, itemsSelectors } from 'state/items';

import { MiddlewareProps } from 'view/shared/RouteMiddlewareController/types';
import useNavigator from 'hook/useNavigator';

const SearchChecker: React.FC<MiddlewareProps> = ({
  next,
}: MiddlewareProps) => {
  const dispatch = useDispatch();
  const { navigateTo, searchParams } = useNavigator();
  const searchStatus = useSelector(itemsSelectors.searchItemsStatus);
  useEffect(() => {
    if (searchStatus === REQUEST_STATUS.FAILED) {
      navigateTo(ROUTES_NAVIGATION.SEARCHER);
    }

    if (searchStatus === REQUEST_STATUS.LOADED) {
      next && next();
    }

    if (searchStatus === REQUEST_STATUS.NOT_STARTED) {
      if (searchParams.includes('?search')) {
        const searchParam = searchParams.replace('?search=', '')
        dispatch(itemsActions.searchItems(searchParam));
      } else {
        navigateTo(ROUTES_NAVIGATION.SEARCHER);
      }
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [searchStatus]);

  return null;
};

export default SearchChecker;
