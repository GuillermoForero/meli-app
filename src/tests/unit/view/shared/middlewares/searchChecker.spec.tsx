import React from 'react';
import { render } from 'tests/support/customRender';
import { Provider } from 'react-redux';
import createMockStore from 'redux-mock-store';
import RouteMiddlewareController from 'view/shared/RouteMiddlewareController';
import { appStoreEmpty, appStoreFull } from 'tests/support/stubAppStore';
import SearchChecker from 'view/shared/middleware/SearchChecker';

const FakeComponent = () => {
  return <p>This is a Fake Component</p>;
};

const mockStore = createMockStore();

describe('[Middleware SearchChecker]', () => {
  it('Runs a setup SearchChecker middleware', () => {
    const store = mockStore(appStoreFull);

    const { container, getByText } = render(
      <Provider store={store}>
        <RouteMiddlewareController middlewares={[<SearchChecker />]}>
          <FakeComponent />
      </RouteMiddlewareController>
      </Provider>
    );

    expect(getByText('This is a Fake Component')).toBeDefined();
    expect(container).toMatchSnapshot();
  });
  it('SearchChecker middleware with empty store', () => {
    const emptyStore = mockStore(appStoreEmpty);

    const { container } = render(
      <Provider store={emptyStore}>
        <RouteMiddlewareController middlewares={[<SearchChecker />]}>
          <FakeComponent />
      </RouteMiddlewareController>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
