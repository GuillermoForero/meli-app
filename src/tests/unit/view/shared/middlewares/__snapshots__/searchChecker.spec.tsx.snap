// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Middleware SearchChecker] Runs a setup SearchChecker middleware 1`] = `
<div>
  <p>
    This is a Fake Component
  </p>
</div>
`;

exports[`[Middleware SearchChecker] SearchChecker middleware with empty store 1`] = `<div />`;
